### PR TITLE
faster Float32 and Float16 pow

### PR DESCRIPTION
**Pull request by oscardssmith**
Returns `Float32` speed to roughly the speed of 1.5. Also speeds up `Float16` pow. These new methods are .5 ULP (from limited testing). There is further room to optimize these, but this fixes the regression. At some point, I hope to have a `Float64` version, but that will be much harder as it requires a `log2` function that gives extra bits of precision, and an `exp2` function that takes in a `Double Double`

### CODE REVIEW

**Comment by oscardssmith** [comment]
Do we require `pow` produce exact results for `Integer` arguments? That's the test that is currently failing.

**Comment by oscardssmith** [comment]
I think this is ready to go. I haven't done fully rigorous tests on it, but every test-case I've used has worked, and conceptually, this should be just over .5 ULP.

**Comment by oscardssmith** [comment]
I've tested a wide variety of random numbers and haven't found more than `.5` ULP. Given that current `pow` is system dependent (about `.75` ULP on my system), I think we should merge.

**Comment by oscardssmith** [comment]
Bumping this. Can someone look at it and merge?

File: base/math.jl [hunk 1]
```diff
@@ -867,29 +867,35 @@ end
     z
 end
 @inline function ^(x::Float32, y::Float32)
-    z = ccall("llvm.pow.f32", llvmcall, Float32, (Float32, Float32), x, y)
+    z = Float32(exp2_fast(log2(Float64(x))*y))
```
Comment (vtjnash) [inline base/math.jl:870]: why don't we want to trust llvm (aka libm) here anymore?

File: base/math.jl [hunk 2]
```diff
@@ -867,29 +867,35 @@ end
...
-@inline ^(x::Float16, y::Float16) = Float16(Float32(x)^Float32(y))  # TODO: optimize
 
 @inline function ^(x::Float64, y::Integer)
     y == -1 && return inv(x)
     y == 0 && return one(x)
     y == 1 && return x
     y == 2 && return x*x
     y == 3 && return x*x*x
-    ccall("llvm.pow.f64", llvmcall, Float64, (Float64, Float64), x, Float64(y))
+    return x^Float64(y)
```
Comment (vtjnash) [inline base/math.jl:890]: Why weren't we using the llvm intrinsic for this (isnan(z) should be impossible, IIUC, so the error checking in the call to `^` isn't necessary)?
```suggestion
    ccall("llvm.powi.f64", llvmcall, Float64, (Float64, Int32), x, clamp(y, Int32))
```

File: base/math.jl (hunk 1 above)
Comment (oscardssmith) [reply base/math.jl:870]: 1.6 introduced about a 3x regression on this due to a switch in which libm got loaded.

File: base/math.jl (hunk 2 above)
Comment (oscardssmith) [reply base/math.jl:890]: `prevfloat(1.0)^(2^33)==0.9999990463261383`, so this isn't a valid definition for large `y`

Comment (vtjnash) [reply base/math.jl:890]: That is a good reason. Apparently "under-the-hood" it also just uses essentially the "power-by-squaring" method, which seems to have somewhat low accuracy there:
```
julia> Base.power_by_squaring(prevfloat(1.0), 2^62)
4.3774743599864475e-223

julia> ^(prevfloat(1.0), 2^62)
4.377491037052927e-223

julia> ^(big(prevfloat(1.0)), 2^62)
4.377491037052927038343183444930184380098189891120230882184167270887044427471645e-223
```

File: base/math.jl (hunk 1 above)
Comment (vtjnash) [reply base/math.jl:870]: Should we fix openlibm?

File: base/math.jl (hunk 2 above)
Comment (oscardssmith) [reply base/math.jl:890]: Yeah. Power by squaring necessarily rounds `log_2(y)` roundings

File: base/math.jl (hunk 1 above)
Comment (oscardssmith) [reply base/math.jl:870]: probably. That said, I generally think that we shouldn't be relying on external algorithms, so I'm not that sad about replacing this anyway.

**Comment by kimikage** [comment]
@oscardssmith, does this cause a problem when used with `@fastmath`?
```julia
julia> versioninfo()
Julia Version 1.7.0-DEV.1006
Commit 248c02f531* (2021-04-24 17:37 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)

julia> Float16(-1)^2
Float16(1.0)

julia> @fastmath Float16(-1)^2
ERROR: DomainError with -1.0:
log2 will only return a complex result if called with a complex argument. Try log2(Complex(x)).
```

**Comment by kimikage** [comment]
Another case (found in ColorVectorSpace.jl)
```julia
julia> (-1.0f0)^2.0f0
ERROR: DomainError with -1.0:
log2 will only return a complex result if called with a complex argument. Try log2(Complex(x)).
```

cf. PkgEval: https://github.com/JuliaCI/NanosoldierReports/blob/master/pkgeval/by_date/2021-04/25/report.md

**Comment by oscardssmith** [comment]
That's unfortunate. I'll revert tomorrow unless I can think of something clever to fix it without too much performance impact.

**Pull request merged**